File: cpp/02_medium/src/Snake.cpp
#include "../include/Snake.h"
#include <random>

Snake::Snake() : color(sf::Color::Green), invincible(false), 
                 speedMultiplier(1.0f), invincibleTimer(0), speedTimer(0) {
    reset();
}

void Snake::reset() {
    body.clear();
    body.push_back({10, 10});
    body.push_back({9, 10});
    body.push_back({8, 10});
    dir = RIGHT;
    nextDir = RIGHT;
    invincible = false;
    speedMultiplier = 1.0f;
    invincibleTimer = 0;
    speedTimer = 0;
}

void Snake::move() {
    // Проверка направления
    if ((dir == UP && nextDir != DOWN) ||
        (dir == DOWN && nextDir != UP) ||
        (dir == LEFT && nextDir != RIGHT) ||
        (dir == RIGHT && nextDir != LEFT)) {
        dir = nextDir;
    }
    
    // Новая голова
    sf::Vector2i head = body.front();
    
    switch (dir) {
        case UP:    head.y--; break;
        case DOWN:  head.y++; break;
        case LEFT:  head.x--; break;
        case RIGHT: head.x++; break;
    }
    
    // Добавление новой головы
    body.push_front(head);
    body.pop_back();
}

void Snake::grow() {
    body.push_back(body.back());
}

void Snake::setDirection(Direction newDir) {
    nextDir = newDir;
}

void Snake::setInvincible(bool inv, int duration) {
    invincible = inv;
    invincibleTimer = duration;
}

void Snake::setSpeedMultiplier(float mult, int duration) {
    speedMultiplier = mult;
    speedTimer = duration;
}

void Snake::updateEffects() {
    if (invincibleTimer > 0) {
        invincibleTimer--;
        if (invincibleTimer <= 0) {
            invincible = false;
        }
    }
    
    if (speedTimer > 0) {
        speedTimer--;
        if (speedTimer <= 0) {
            speedMultiplier = 1.0f;
        }
    }
}

bool Snake::checkSelfCollision() const {
    if (invincible) return false;
    
    auto head = body.front();
    for (size_t i = 1; i < body.size(); i++) {
        if (body[i] == head) {
            return true;
        }
    }
    return false;
}

bool Snake::checkPosition(const sf::Vector2i& pos) const {
    for (const auto& segment : body) {
        if (segment == pos) {
            return true;
        }
    }
    return false;
}

void Snake::draw(sf::RenderWindow& window, int cellSize) const {
    for (size_t i = 0; i < body.size(); i++) {
        sf::RectangleShape rect(sf::Vector2f(cellSize - 2, cellSize - 2));
        rect.setPosition(body[i].x * cellSize + 1, body[i].y * cellSize + 1);
        
        if (i == 0) {
            // Голова
            if (invincible) {
                float pulse = (sinf(invincibleTimer * 0.1f) + 1.0f) * 0.5f;
                rect.setFillColor(sf::Color(255, 255, 0, 255 * pulse));
            } else {
                rect.setFillColor(sf::Color(50, 255, 50));
            }
            rect.setOutlineColor(sf::Color::White);
            rect.setOutlineThickness(1);
        } else {
            // Тело
            float opacity = 1.0f - (i * 0.05f);
            if (opacity < 0.3f) opacity = 0.3f;
            rect.setFillColor(sf::Color(0, 200, 0, 255 * opacity));
        }
        
        window.draw(rect);
        
        // Глаза для головы
        if (i == 0) {
            sf::CircleShape eye(2);
            eye.setFillColor(sf::Color::White);
            
            switch (dir) {
                case RIGHT:
                    eye.setPosition(body[i].x * cellSize + cellSize - 8, 
                                   body[i].y * cellSize + 5);
                    window.draw(eye);
                    eye.setPosition(body[i].x * cellSize + cellSize - 8, 
                                   body[i].y * cellSize + cellSize - 9);
                    window.draw(eye);
                    break;
                case LEFT:
                    eye.setPosition(body[i].x * cellSize + 4, 
                                   body[i].y * cellSize + 5);
                    window.draw(eye);
                    eye.setPosition(body[i].x * cellSize + 4, 
                                   body[i].y * cellSize + cellSize - 9);
                    window.draw(eye);
                    break;
                case UP:
                    eye.setPosition(body[i].x * cellSize + 5, 
                                   body[i].y * cellSize + 4);
                    window.draw(eye);
                    eye.setPosition(body[i].x * cellSize + cellSize - 9, 
                                   body[i].y * cellSize + 4);
                    window.draw(eye);
                    break;
                case DOWN:
                    eye.setPosition(body[i].x * cellSize + 5, 
                                   body[i].y * cellSize + cellSize - 8);
                    window.draw(eye);
                    eye.setPosition(body[i].x * cellSize + cellSize - 9, 
                                   body[i].y * cellSize + cellSize - 8);
                    window.draw(eye);
                    break;
            }
        }
    }
}

File: csharp/02_medium/Snake.cs
using System;
using System.Collections.Generic;
using System.Drawing;

namespace SnakeGameWinForms
{
    public class Snake
    {
        private LinkedList<Point> body;
        private Direction direction;
        private Direction nextDirection;
        private bool growFlag;

        public Snake()
        {
            body = new LinkedList<Point>();
            Reset();
        }

        public void Reset()
        {
            body.Clear();
            body.AddFirst(new Point(10, 10));
            body.AddFirst(new Point(9, 10));
            body.AddFirst(new Point(8, 10));
            direction = Direction.Right;
            nextDirection = Direction.Right;
            growFlag = false;
        }

        public void Move()
        {
            // Проверка направления
            if ((direction == Direction.Up && nextDirection != Direction.Down) ||
                (direction == Direction.Down && nextDirection != Direction.Up) ||
                (direction == Direction.Left && nextDirection != Direction.Right) ||
                (direction == Direction.Right && nextDirection != Direction.Left))
            {
                direction = nextDirection;
            }

            // Новая голова
            Point head = body.First.Value;

            switch (direction)
            {
                case Direction.Up: head.Y--; break;
                case Direction.Down: head.Y++; break;
                case Direction.Left: head.X--; break;
                case Direction.Right: head.X++; break;
            }

            // Добавление новой головы
            body.AddFirst(head);

            // Удаление хвоста
            if (!growFlag)
            {
                body.RemoveLast();
            }
            else
            {
                growFlag = false;
            }
        }

        public void Grow()
        {
            growFlag = true;
        }

        public void ChangeDirection(Direction newDir)
        {
            nextDirection = newDir;
        }

        public Point GetHead()
        {
            return body.First.Value;
        }

        public List<Point> GetBody()
        {
            return new List<Point>(body);
        }

        public bool CheckSelfCollision()
        {
            var head = body.First.Value;
            var current = body.First.Next;

            while (current != null)
            {
                if (current.Value == head)
                {
                    return true;
                }
                current = current.Next;
            }

            return false;
        }

        public void Draw(Graphics g, int cellSize)
        {
            int index = 0;
            foreach (var point in body)
            {
                Rectangle rect = new Rectangle(
                    point.X * cellSize + 1,
                    point.Y * cellSize + 1,
                    cellSize - 2,
                    cellSize - 2
                );

                if (index == 0)
                {
                    // Голова
                    using (SolidBrush brush = new SolidBrush(Color.FromArgb(50, 255, 50)))
                    {
                        g.FillRectangle(brush, rect);
                    }

                    // Глаза
                    using (SolidBrush eyeBrush = new SolidBrush(Color.White))
                    {
                        switch (direction)
                        {
                            case Direction.Right:
                                g.FillEllipse(eyeBrush, rect.X + cellSize - 8, rect.Y + 3, 4, 4);
                                g.FillEllipse(eyeBrush, rect.X + cellSize - 8, rect.Y + cellSize - 7, 4, 4);
                                break;
                            case Direction.Left:
                                g.FillEllipse(eyeBrush, rect.X + 4, rect.Y + 3, 4, 4);
                                g.FillEllipse(eyeBrush, rect.X + 4, rect.Y + cellSize - 7, 4, 4);
                                break;
                            case Direction.Up:
                                g.FillEllipse(eyeBrush, rect.X + 3, rect.Y + 4, 4, 4);
                                g.FillEllipse(eyeBrush, rect.X + cellSize - 7, rect.Y + 4, 4, 4);
                                break;
                            case Direction.Down:
                                g.FillEllipse(eyeBrush, rect.X + 3, rect.Y + cellSize - 8, 4, 4);
                                g.FillEllipse(eyeBrush, rect.X + cellSize - 7, rect.Y + cellSize - 8, 4, 4);
                                break;
                        }
                    }
                }
                else
                {
                    // Тело
                    int alpha = 255 - (index * 15);
                    if (alpha < 100) alpha = 100;
                    using (SolidBrush brush = new SolidBrush(Color.FromArgb(alpha, 0, 200, 0)))
                    {
                        g.FillRectangle(brush, rect);
                    }
                }
                index++;
            }
        }
    }
}
